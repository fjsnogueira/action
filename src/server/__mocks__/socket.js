const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
};

export default socket;
