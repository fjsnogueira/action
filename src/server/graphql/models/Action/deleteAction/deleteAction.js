import getRethink from 'server/database/rethinkDriver';
import {
  GraphQLNonNull,
  GraphQLBoolean,
  GraphQLID
} from 'graphql';
import {requireSUOrTeamMember, requireWebsocket} from 'server/utils/authorization';

export default {
  type: GraphQLBoolean,
  description: 'Delete an action',
  args: {
    actionId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The actionId (teamId::shortid) to delete'
    }
  },
  async resolve(source, {actionId}, {authToken, socket}) {
    const r = getRethink();

    // AUTH
    const [teamId] = actionId.split('::');
    requireSUOrTeamMember(authToken, teamId);
    requireWebsocket(socket);

    // RESOLUTION
    await r.table('Action').get(actionId).delete();
  }
};
