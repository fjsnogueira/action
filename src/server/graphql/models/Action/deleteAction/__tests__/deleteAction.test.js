import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import deleteAction from 'server/graphql/models/Action/deleteAction/deleteAction';

MockDate.set(__now);
console.error = jest.fn();

describe('deleteAction', () => {
  test('deletes an action', async() => {
    // SETUP
    const r = getRethink();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();

    // TEST
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);
    await deleteAction.resolve(undefined, {actionId}, {authToken, socket});

    // VERIFY
    const dbAction = await r.table('Action').get(actionId);
    expect(dbAction).toBe(null);
  });

  test('throws when no websocket is present', async() => {
    const authToken = {};
    await expectAsyncToThrow(deleteAction.resolve(undefined, {actionId: ''}, {authToken}));
  });

  test('throw when the caller is not a team member', async() => {
    // SETUP
    const mockDB = new MockDB();
    const {user} = await mockDB.init();

    // TEST
    const authToken = mockAuthToken(user[1]);
    const actionId = 'foo::bar';
    await expectAsyncToThrow(deleteAction.resolve(undefined, {actionId}, {authToken, socket}), [mockDB.context.team.id]);
  });
});
