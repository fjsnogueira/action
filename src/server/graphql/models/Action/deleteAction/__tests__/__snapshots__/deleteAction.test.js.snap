// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteAction throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to team foo\\"}"`;

exports[`deleteAction throws when no websocket is present 1`] = `"{\\"_error\\":\\"You do not have access to team \\"}"`;
